loader/compat_php.c

```diff
@@ -0,0 +1,91 @@
+#include <php.h>
+#include <stdbool.h>
+
+#include "compat_php.h"
+
+#define HT_FLAGS(ht) (ht)->u.flags
+#define PHP_70_71_72_IS_STR_INTERNED (1<<1)
+#define PHP_70_71_72_ZSTR_IS_INTERNED(s) (GC_FLAGS(s) & PHP_70_71_72_IS_STR_INTERNED)
+
+ZEND_API zval* ZEND_FASTCALL zend_hash_set_bucket_key(HashTable *ht, Bucket *b, zend_string *key) __attribute__((weak));
+
+static bool ddloader_zstr_is_interned(int php_api_no, zend_string *key) {
+    if (php_api_no <= 20170718) {  // PHP 7.0 - 7.2
+        return PHP_70_71_72_ZSTR_IS_INTERNED(key);
+    } else {
+        return ZSTR_IS_INTERNED(key);
+    }
+}
+
+// This is an adaptation of zend_hash_set_bucket_key which is only available only starting from PHP 7.4
+// to be compatible with PHP 7.0+
+zval* ddloader_zend_hash_set_bucket_key(int php_api_no, HashTable *ht, Bucket *b, zend_string *key) {
+    // Use the real implementation if it exists
+    if (zend_hash_set_bucket_key) {
+        return zend_hash_set_bucket_key(ht, b, key);
+    }
+
+    // Fallback for PHP < 7.4
+	uint32_t nIndex;
+	uint32_t idx, i;
+	Bucket *p, *arData;
+
+	ZEND_ASSERT(!(HT_FLAGS(ht) & HASH_FLAG_PACKED));
+
+	p = (Bucket*)zend_hash_find(ht, key);
+	if (UNEXPECTED(p)) {
+		return (p == b) ? &p->val : NULL;
+	}
+
+	if (!ddloader_zstr_is_interned(php_api_no, key)) {
+		zend_string_addref(key);
```

Comment:
zend_string_addref() also does to an interned check.
Can we please do?
```suggestion
		GC_ADDREF(key);
```

using
```
#if PHP_VERSION_ID < 70300
#define GC_ADDREF(x) (++GC_REFCOUNT(x))
#define GC_DELREF(x) (--GC_REFCOUNT(x))
#endif
```

And later, in ddloader_zend_hash_set_bucket_key, instead of `if (php_api_no > 20170718) {` also do `if (!ddloader_zstr_is_interned(php_api_no, b->key)) { if (GC_DELREF(b->key) == 0) pefree(s, GC_FLAGS(s) & (php_api_no <= 20170718 ? 1 : IS_STR_PERSISTENT)); }`? Instead of this _leak_ thing?

Reply:
Ideally, we'd just have api_no-aware `ddloader_zend_string_addref()`, `ddloader_zend_string_init()` and `ddloader_zend_string_init()`, with proper persistent strings too. I don't feel too comfortable with using non-persistent strings in the minit there.

Reply:
Fixed.

FI, I haven't put in the code these defines"
```
#if PHP_VERSION_ID < 70300
#define GC_ADDREF(x) (++GC_REFCOUNT(x))
#define GC_DELREF(x) (--GC_REFCOUNT(x))
#endif
```

IMHO, it's just fine to say that this extension can be compiled only on PHP >= 7.4. I added a README.md to mention that.

Reply:
Okay, I'm fine with that.